src/sage/rings/function_field/drinfeld_modules/finite_drinfeld_module.py

```diff
@@ -248,22 +250,177 @@ def frobenius_charpoly(self, var='X'):
 
         ALGORITHM:
 
-            We compute the Frobenius norm, and with it the Frobenius
-            trace. This gives the Frobenius characteristic polynomial.
-            See [MS2019]_, Section 4.
+            This wrapper dispatches the characteristic polynomial
+            computation to the method implementing the corresponding
+            algorithm. By default chooses the 'crystalline' algorithm.
+            See the corresponding docstring for each method for details.
+        """
+        algorithms = {'gekeler', 'crystalline'}
+        if algorithm in algorithms:
+            return getattr(self, \
+                    f'{self.frobenius_charpoly.__name__}_{algorithm}')(var)
+        raise NotImplementedError(f'Algorithm \"{algorithm}\" not implemented')
+
+    def frobenius_charpoly_gekeler(self, var='X'):
+        r"""
+        Return the characteristic polynomial of the Frobenius
+        endomorphism for any rank if the minimal polynomial is
+        equal to the characteristic polynomial. Currently only
+        works for Drinfeld modules defined over Fq[T].
+
+        WARNING: This algorithm only works in the "generic" case
+                 when the corresponding linear system is invertible.
+                 Notable cases where this fails include Drinfeld
+                 modules whose minimal polynomial is not equal to
+                 the characteristic polynomial, and rank 2 Drinfeld
+                 modules where the degree 1 coefficient of \phi_T is
+                 0.
+
+        INPUT:
+
+        - ``var`` (default: ``'X'``) -- the name of the second variable
```

Comment:
```suggestion
        - ``var`` (string, default: ``'X'``) -- the name of the second variable
```

Also, I would change `var` to `name`.